Allow blocks in deployment params

GalLadislav opened the pull request:
### Context

Related #14393

I tried to "fix" Prefect's 3.0 inability to pass Blocks as deployments parameters. This feature is really important for me personally (and probably for others too). Since the related issue was closed as "not planned", i implemented something that could provide at least some functionality until it is properly addressed. 

I would appreciate a feedback if this PR is useful and is worthwhile to polish it and add unittests etc. Thank you in advance for any guidance.

### Implementation

Implementation is building on already existing UI features in FlowRunCreate for deployment runs that use passing `{"$ref": "__uuid__"}` as block reference. I have tracked to main issues which causes this feature to not work. 
1. Server's validation API does not allow this block definition,
2. `Block` does not processes the **$ref** in init kwargs.

#### Server validation API

To allow **$ref** for blocks, preprocessing schemas modify block definition to allow either block definition or block reference.
Added two **$ref** formats, one that supports deployment flow run create form and one that supports multiple reference types.

#### Block initialization

For Block class i added method `Block.load_from_ref` that process either block document ID or supported reference dictionary format and returns loaded block with data from references block document.
Other changes:
- created `Block._load_from_block_document` method to deal with duplicate code used both in `Block.load` and `Block.load_from_ref`.
- added `Block._get_block_document_by_id`
- `Flow.validate_parameters` updated to transform any block reference in given parameters to corresponding block with `Block.load_from_ref`.

#### Issues

- API validation might check if blocks exist

### Checklist

- [x] This pull request references any related issue by including "closes `<link to issue>`.
- [x] If this pull request adds new functionality, it includes unit tests that cover the changes
- [x] If this pull request adds functions or classes, it includes helpful docstrings.

cicdw commented on src/prefect/blocks/core.py:
```diff
@@ -790,6 +800,33 @@ async def _get_block_document(
+    @inject_client
+    async def _get_block_document_by_id(
+        cls,
+        block_document_id: str | uuid.UUID,
```
we are still supporting Python 3.9 and so can't support this syntax quite yet! You'll need to use `typing.Union` here

GalLadislav replied:
Thank you, fixed

cicdw commented on src/prefect/blocks/core.py:
```diff
@@ -790,6 +800,33 @@ async def _get_block_document(
+            except ValueError as e:
+                raise ValueError(
+                    f"Block document ID {block_document_id!r} is not a valid UUID"
+                ) from e
```
```suggestion
                )
```
no need to add more to the traceback here

cicdw commented on src/prefect/blocks/core.py:
```diff
@@ -790,6 +800,33 @@ async def _get_block_document(
+        except prefect.exceptions.ObjectNotFound as e:
+            raise ValueError(
+                f"Unable to find block document with ID {block_document_id!r}"
+            ) from e
```
```suggestion
            )
```

cicdw commented on src/prefect/blocks/core.py:
```diff
@@ -272,6 +273,15 @@ class Block(BaseModel, ABC):
     def __init__(self, *args, **kwargs):
+        block_document_id = kwargs.pop("$ref", None)
+        if block_document_id:
+            block_document, block_document_name = self._get_block_document_by_id(
```
because this is _always_ run synchronously, I suggest adding `_sync=True` to this call to ensure it's called properly.

cicdw reviewed (changes requested):
a few code nit picks - also could you add a test for what this initialization looks like, and also a test for using a block type hint with a flow that validates its parameters?

Ultimately it'd be better (and easier to reason about) if the API could pass the block slug to the client for validation, but this might be an OK patch in the meantime as that will involve more changes.

I've also tagged @desertaxle for review here, as he might have more insight into this implementation and whether there are any edge cases we should consider.

GalLadislav commented:
I added block reference validation to check if it matches initialized block type. Not sure if both slug and type should be compared, or only one of them or should be validated differently.
Added also some tests to check basic cases. Probably more should be added. Also I was not sure if the location is ok.

desertaxle commented on src/prefect/utilities/schema_tools/validation.py:
```diff
@@ -253,5 +253,24 @@ def preprocess_schema(
                 process_properties(
                     definition["properties"], required_fields, allow_none_with_default
                 )
+            # Allow block types to be referenced by their id
+            if "block_type_slug" in definition:
+                schema["definitions"][definition["title"]] = {
+                    "oneOf": [
+                        definition,
+                        {
+                            "type": "object",
+                            "properties": {
+                                "$ref": {
+                                    "type": "string",
+                                    "format": "uuid",
+                                },
+                            },
+                            "required": [
+                                "$ref",
+                            ],
+                        },
+                    ]
+                }
```
The reference format used when nested block documents in other block documents are `{"$ref": {"block_document_id": block_document_id } }.` I'd prefer to use that same format here for consistency and to leave room for other reference types in the future (like using the block slug as @cicdw suggested).

GalLadislav replied:
Do i understand correctly that you request changes to the UI itself? i tried to fit current solution to current UI functions, since changes to UI would require more work. I can do that, but it would take more time.

GalLadislav replied:
Added additional schema to enable your proposed reference formats. I kept the current format, as it is used by Deployment Flow Run Create form. I hope i understood your request correctly.

desertaxle replied:
Thanks for adding that! I forgot that we'd need to update the UI to populate the backend data to match that format. I can create a separate issue to allow different reference shapes.

desertaxle commented on src/prefect/blocks/core.py:
```diff
@@ -272,6 +273,19 @@ class Block(BaseModel, ABC):
     def __init__(self, *args, **kwargs):
+        block_document_id = kwargs.pop("$ref", None)
+        if block_document_id:
+            block_document, block_document_name = self._get_block_document_by_id(
+                block_document_id, _sync=True
+            )
```
I'm wary of putting a blocking API call in `__init__`. Could you add a new method to load a block from an ID? That would require an update to `Flow.validate_parameters` to use that new method, but I think that'd make this new functionality more explicit.

cicdw replied:
Good call, agreed on this -- maybe it could be a classmethod `load_from_ref` that the `validate_parameters` method calls explicitly for Block typed parameters.

GalLadislav replied:
Refactored to new method `load_from_ref` which is used for preprocessing parameters in `validate_parameters`. To deal with duplicate code i dared to create `_load_from_block_document` method to be used both by `load` and by `load_from_ref`.

desertaxle reviewed (changes requested):
Thanks for the PR @GalLadislav, this will be a super useful feature! Let me know if you have any questions about any of my requested changes!

cicdw commented on tests/blocks/test_block_reference.py:
```diff
@@ -0,0 +1,124 @@
+
+    class OtherReferencedBlock(Block):
+        c: int
+        d: str
```
Could you write a test for `load_from_ref` that includes a nested block document? e.g., 
```python
class NestedBlock(Block):
    child: OtherReferencedBlock
```

GalLadislav replied:
I am glad you noticed that these are missing. Added few more test for this including nesting with `pydantic.BaseModel` as flow parameter.

cicdw reviewed (changes requested):
This looks pretty solid to me! I requested one additional test for nested block references, and will wait on @desertaxle's final review as well. 

Thank you for taking the time to work on this @GalLadislav !!

desertaxle reviewed (approved):
LGTM! Thanks for the awesome contribution @GalLadislav!

cicdw reviewed (approved):
Thanks again @GalLadislav for working with us on this and being so responsive to our feedback!!

GalLadislav commented:
Thank you @desertaxle @cicdw for your time and guidance. I am happy i could contribute.
